src/iterator.c

```diff
@@ -1587,7 +2023,7 @@ bool git_iterator_current_is_ignored(git_iterator *iter)
 {
 	workdir_iterator *wi = (workdir_iterator *)iter;
 
-	if (iter->type != GIT_ITERATOR_TYPE_WORKDIR)
+	if (!GIT_ITERATOR_TYPE_IS_WORKDIR_OR_FILELIST(iter->type))
```

Comment:
Just found several places where we do this.  Cast "iter" into "wi" before we know if it of the correct type.  This exists in the code I added, but also in the original code.

I'm fixing the instances I introduced, but the overall pattern seems dangerous.